### PR TITLE
Fix type extraction with asymetric operators 

### PR DESCRIPTION
**Pull request by tobim**

### CODE REVIEW

File: libvast/src/expression_visitors.cpp [hunk 1]
```diff
@@ -332,25 +332,27 @@ expected<expression> type_resolver::operator()(const predicate& p) {
...
+    connective.emplace_back(predicate{std::move(x), op_, d});
   }
-  if (dis.empty())
+  if (connective.empty())
     return expression{}; // did not resolve
-  else if (dis.size() == 1)
-    return {std::move(dis[0])};
+  if (connective.size() == 1)
+    return {std::move(connective[0])};
+  if (op_ == not_in || op_ == not_match)
```
Comment (mavam) [inline libvast/src/expression_visitors.cpp]:
```suggestion
  if (op_ == not_equal || op_ == not_match || op_ == not_in || op_ == not_ni)
```

We might not trigger all clauses of this disjunction, but logically, these are I think all cases that require De Morgan.

Comment (tobim) [reply libvast/src/expression_visitors.cpp]: thanks

**Comment by mavam** [comment]
How about a changelog entry for this one? It is a user-facing bugfix after all.

**Comment by mavam** [comment]
It just came to my mind that the type extractors are not the only case where this is a problem. What about key extrators? IIRC, if we have a query like `orig_h == 6.6.6.6`, then the index will go over all types where `orig_h` is valid suffix (e.g., `conn`, `http`, `smtp`, etc.) and also make a disjunction of them.

Would you mind double-checking that this bug doesn't apply to other extractor types as well?

**Comment by tobim** [comment]
> It just came to my mind that the type extractors are not the only case where this is a problem. What about key extrators? IIRC, if we have a query like `orig_h == 6.6.6.6`, then the index will go over all types where `orig_h` is valid suffix (e.g., `conn`, `http`, `smtp`, etc.) and also make a disjunction of them.
> 
> Would you mind double-checking that this bug doesn't apply to other extractor types as well?

I don't think I understand what you are saying. If we have multiple types with the same suffix they are usually in different tables, like your examples with `conn`, `http`, `smtp`. If we do have a record with the structure `record{ orig.host: addr, resp.host: addr }`, then we might have a problem there. Is that what you mean?

**Comment by tobim** [comment]
So for key extractors this is problem does exist as shown by the newly added tests. Commit  
https://github.com/vast-io/vast/pull/460/commits/9260ea0256a33f6bc2799cd59afb0766b405711e also contains the fix.

**Pull request merged**